src/mlir/generator.cc

```diff
@@ -93,21 +93,82 @@ namespace mlir::verona
   llvm::Error Generator::parseModule(const ::ast::Ast& ast)
   {
     assert(AST::isClass(ast) && "Bad node");
+    assert(AST::getID(ast) == "$module" && "Bad node");
     module = mlir::ModuleOp::create(getLocation(ast));
-    // TODO: Support more than just functions at the module level
+
     auto body = AST::getClassBody(ast);
-    llvm::SmallVector<::ast::WeakAst, 4> funcs;
-    AST::getSubNodes(funcs, body);
-    for (auto f : funcs)
+    llvm::SmallVector<::ast::WeakAst, 4> globals;
+    AST::getSubNodes(globals, body);
+    for (auto g : globals)
     {
-      auto fun = parseFunction(f.lock());
-      if (auto err = fun.takeError())
-        return err;
-      module->push_back(*fun);
+      if (AST::isClass(g))
+      {
+        auto c = parseClass(g.lock());
+        if (auto err = c.takeError())
+          return err;
+        module->push_back(*c);
+      }
+      else if (AST::isFunction(g))
+      {
+        auto f = parseFunction(g.lock());
+        if (auto err = f.takeError())
+          return err;
+        module->push_back(*f);
+      }
     }
     return llvm::Error::success();
   }
 
+  llvm::Expected<mlir::verona::ClassOp>
+  Generator::parseClass(const ::ast::Ast& ast)
+  {
+    assert(AST::isClass(ast) && "Bad node");
+
+    // Add the type first
+    // Declare before building fields to allow for recursive declaration
+    auto name = AST::getID(ast);
+    auto type = ClassType::get(context, name);
+    typeTable.insert(name, type);
+
+    // Field names and types
+    llvm::SmallVector<::ast::WeakAst, 1> names;
+    AST::getSubNodes(names, AST::getClassBody(ast));
```

Comment:
Sorry isn't the following equivalent.  It doesn't require a zip or a method call that returns the types of sub nodes, and then reconstructs their pairing?

```C++
llvm::SmallVector<::ast::WeakAst, 1> ast_fields;
AST::getSubNodes(ast_fields, AST::getClassBody(ast));
llvm::SmallVector<std::pair<StringRef, mlir::Type>, 4> fields;
for (auto field : ast_fields)
{
   auto field_name = AST::getID(field);
   auto field_type = parseType(AST::getType(field).lock());
   fields.push_back({fieldName, fieldType});
}
```

Reply:
This is also much closer to what you do below in this function with the builder?

Reply:
Ah, yes. This is indeed, simpler.

My aim on creating them was to make sure all my expectations for building the MLIR nodes from AST were as explicitly as possible, even if that meant duplicating logic. So I took the task of exposing every "semantic extraction" as a new method, but I think your proposal is indeed simpler and still "semantically" equivalent.

The reason why I did this is first to make sure I got the right semantics, and second to clearly see what methods are the same and what are not. I guess it's already working. :)